Workload partitioning

lack opened the pull request:
This series of commits:
- Adds github.com/cri-o/cpuset
- Adds a new install-config.yaml 'workloadSettings' section (optional) as outlined in https://github.com/openshift/enhancements/pull/703, which uses the cpuset library to validate the input
- Uses the new 'workloadSettings' section to generate MachineConfig objects that configure both Cri-o and Kubelet on installed hosts to use the specified cpuset for management workloads.

lack commented:
/cc @staebler

staebler commented on pkg/types/validation/installconfig.go:
```diff
@@ -609,3 +611,29 @@ func validateIPProxy(proxy string, n *types.Networking, fldPath *field.Path) fie
+			allErrs = append(allErrs, field.Duplicate(partitionFldPath.Child("name"), p.Name))
+		}
+		partitionNames[p.Name] = true
+		if p.CPUIds == "" {
+			allErrs = append(allErrs, field.Invalid(partitionFldPath.Child("cpuIDs"), p.CPUIds, "cannot be empty"))
+		} else if cpuset, err := cpuset.Parse(p.CPUIds); err != nil || cpuset.IsEmpty() {
+			// Re-check if the parsed cpuset is empty, since the parser may interpret some errors as the empty set.
+			allErrs = append(allErrs, field.Invalid(partitionFldPath.Child("cpuIDs"), p.CPUIds, "could not parse the cpuset"))
+		}
```
Do we need to check for empty twice? If the `IsEmpty` check sufficient?
```suggestion
		swtich cpuset, err := cpuset.Parse(p.CPUIds); {
		case err != nil:
			// invalid error
		case cpuset.IsEmpty():
			// required error
		}
```

staebler commented on pkg/types/validation/installconfig.go:
```diff
@@ -609,3 +611,29 @@ func validateIPProxy(proxy string, n *types.Networking, fldPath *field.Path) fie
+		partitionNames[p.Name] = true
+		if p.CPUIds == "" {
+			allErrs = append(allErrs, field.Invalid(partitionFldPath.Child("cpuIDs"), p.CPUIds, "cannot be empty"))
+		} else if cpuset, err := cpuset.Parse(p.CPUIds); err != nil || cpuset.IsEmpty() {
```
The `cpuset` variable name hides the `cpuset` package name.

staebler commented on pkg/types/validation/installconfig_test.go:
```diff
@@ -203,12 +203,23 @@ func validOvirtPlatform() *ovirt.Platform {
+	return c
+}
+
+type testCase struct {
```
Don't add this. Keep the test case type anonymous.

mrunalp commented on pkg/asset/machines/machineconfig/workload-partitioning.go:
```diff
@@ -0,0 +1,127 @@
+		annotation_prefix = "io.openshift.workload.management"
+		resources = { "cpu" = "", "cpuset" = "0-1", }
+	*/
+	crioIni := ini.Empty()
```
The format we use is toml. CRI-O uses "github.com/BurntSushi/toml" to read it.

lack replied:
Looked a lot like ini, but sure, I can switch over :)

lack replied:
@mrunalp : I have a pending change that switches over to `github.com/BurtSushi/toml` and the encoder wants to create a file with this contents... Is it okay?  It looks a bit different than I was expecting:
```
["crio.runtime.workloads.master"]
label = "master.workload.openshift.io/cores"
annotation_prefix = "io.openshift.workload.master"
resources = "{ \"cpu\" = \"\", \"cpuset\" = \"0-1\", }"
```

lack replied:
@mrunalp : A slack conversation with @haircommander revealed that the format produced by `github.com/BurntSushi/toml` doesn't work, but the format produced in this current PR by `gopkg.in/ini.v1` works just fine:
```
[crio.runtime.workloads.management]
label             = "management.workload.openshift.io/cores"
annotation_prefix = "io.openshift.workload.management"
resources         = { "cpu" = "", "cpuset" = "0-1", }
```
So I think it's best to stick with what I've got here - `gopkg.in/ini.v1` is already used elsewhere in this repo anyway.

staebler commented on pkg/types/validation/installconfig_test.go:
```diff
@@ -1208,6 +1219,58 @@ func TestValidateInstallConfig(t *testing.T) {
+			expectedError: `^workloadSettings\[0\].cpuIDs: Invalid value: "": cannot be empty`,
+		},
+	}
+	for _, id := range []string{"0", "0-1", "3,5,7", "3,5-8,11", "2-3,5-6,8-9"} {
+		cases = append(cases, validCpuidTestcase(id))
+	}
+	for _, id := range []string{"nodigits", "-7", "10-", ",2", "4,", "5,,6", "8--10", "2,4,bad", "5,6-,8", "2-3,-6,8-9", "10-6", "0,5-2,7"} {
+		cases = append(cases, invalidCpuidTestcase(id))
 	}
```
These cases are testing the `cpuset.Parse` function, which is not the unit under test.

lack replied:
This was a hold-over from an earlier invocation that used a regex for validation instead of the cpuset module like the current implementation does.  I left it in as proof that the other library does what we want and external change there doesn't break is in the future.  But I suppose the unit test is there in vendors/ if we need it.

Gone.

staebler commented on pkg/asset/machines/machineconfig/workload-partitioning.go:
```diff
@@ -0,0 +1,127 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/installer/pkg/asset/ignition"
+	"github.com/pkg/errors"
```
Move this import to the previous block.

staebler commented on pkg/asset/machines/machineconfig/workload-partitioning.go:
```diff
@@ -0,0 +1,127 @@
+// annotation_prefix = "io.openshift.workload.management"
+// resources = { "cpu" = "", "cpuset" = "0-1", }
+func CrioWorkloadDropinContents(name, cpuset string) (string, error) {
+	/*
+		[crio.runtime.workloads.management]
+		label = "management.workload.openshift.io/cores"
+		annotation_prefix = "io.openshift.workload.management"
+		resources = { "cpu" = "", "cpuset" = "0-1", }
+	*/
```
Remove this. It is the same example in the description of the function.

staebler commented on pkg/asset/machines/machineconfig/workload-partitioning.go:
```diff
@@ -0,0 +1,127 @@
+		Resources:        fmt.Sprintf(`{ "cpu" = "", "cpuset" = "%s", }`, cpuset),
+	})
+	if err != nil {
+		return "", errors.Wrap(err, "Could not reflect structure to INI")
```
Errors should start with lowercase.
```suggestion
		return "", errors.Wrap(err, "could not reflect structure to INI")
```

staebler commented on pkg/asset/machines/machineconfig/workload-partitioning.go:
```diff
@@ -0,0 +1,127 @@
+	Resources        string `ini:"resources"`
+}
+
+// KubeletWorkloadDropinContents generates te content expected by Kubelet for workload partitioning
```
```suggestion
// KubeletWorkloadDropinContents generates the content expected by Kubelet for workload partitioning
```

staebler commented on pkg/asset/machines/machineconfig/workload-partitioning.go:
```diff
@@ -0,0 +1,127 @@
+					fmt.Sprintf("/etc/crio/crio.conf.d/01-%s-workload", name),
+					"root", 0644, crioCfg),
+				ignition.FileFromString(
+					"/etc/kubernetes/workload-pinning",
```
Does this filename not need specificity for the workload? Like `/etc/kubernetes/workload-management-pinning`.

lack replied:
No, this is a single file which contains all the workloads.

staebler commented on pkg/asset/machines/machineconfig/workload-partitioning.go:
```diff
@@ -0,0 +1,127 @@
+			Kind:       "MachineConfig",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("02-%s-workload-partitioning", role),
```
It seems like this name should be specific to the workload as well.

staebler commented on pkg/asset/machines/machineconfig/workload-partitioning.go:
```diff
@@ -0,0 +1,127 @@
+			Kind:       "MachineConfig",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("02-%s-workload-partitioning", role),
```
What is the significance of the "02-" at the start of the machineconfig name?

lack replied:
MCO creates `01-master-container-runtime` and `01-master-kubelet` so it felt like this, which modifies both CRI-O and Kubelet would be +1 from there.  Though strictly speaking I don't think it matters.

staebler commented on pkg/asset/machines/machineconfig/workload-partitioning_test.go:
```diff
@@ -0,0 +1,53 @@
+		err = json.Unmarshal(result.Spec.Config.Raw, &cfg)
+		assert.Equal(t, err, nil, "No err")
+		files := cfg.Storage.Files
+		assert.Equal(t, 2, len(files), "Two files in the machineconfig")
```
```suggestion
		assert.Len(t, files, 2, "Two files in the machineconfig")
```

staebler commented on pkg/asset/machines/master.go:
```diff
@@ -402,6 +402,14 @@ func (m *Master) Generate(dependencies asset.Parents) error {
 		}
 		machineConfigs = append(machineConfigs, ignFIPS)
 	}
+	if len(ic.Workload) > 0 {
+		w := ic.Workload[0]
+		ignWorkload, err := machineconfig.ForWorkloadPartitioning(string(w.Name), w.CPUIds, "master")
+		if err != nil {
+			return errors.Wrap(err, "failed to create ignition for workload partitioning for master machines")
+		}
+		machineConfigs = append(machineConfigs, ignWorkload)
+	}
```
Do this for all workload partitions now so that we don't have to change this later when we support more.
```suggestion
	for _, w := range ic.Workload {
		ignWorkload, err := machineconfig.ForWorkloadPartitioning(string(w.Name), w.CPUIds, "master")
		if err != nil {
			return errors.Wrap(err, "failed to create ignition for workload partitioning for master machines")
		}
		machineConfigs = append(machineConfigs, ignWorkload)
	}
```

lack replied:
Can do.

lack replied:
Turns out it works better to combine all of the partitions into a single machineconfig, since bot the CRI-O and Kubelet config file drop-ins are designed to handle multiple partitions in each, so I did it that way.  Push with the changes coming soon.

staebler commented on pkg/asset/machines/master.go:
```diff
@@ -402,6 +402,14 @@ func (m *Master) Generate(dependencies asset.Parents) error {
+		w := ic.Workload[0]
+		ignWorkload, err := machineconfig.ForWorkloadPartitioning(string(w.Name), w.CPUIds, "master")
+		if err != nil {
+			return errors.Wrap(err, "failed to create ignition for workload partitioning for master machines")
```
Add context for which partition had problems.
```suggestion
			return errors.Wrapf(err, "failed to create ignition for workload partitioning %q for master machines", w.Name)
```

staebler commented on pkg/asset/machines/master_test.go:
```diff
@@ -10,16 +11,57 @@ import (
 	"github.com/openshift/installer/pkg/asset/rhcos"
 	"github.com/openshift/installer/pkg/types"
 	awstypes "github.com/openshift/installer/pkg/types/aws"
+	"github.com/vincent-petithory/dataurl"
```
Move this to the previous block of imports.

staebler commented on pkg/asset/machines/master_test.go:
```diff
@@ -10,16 +11,57 @@ import (
 		name                  string
 		key                   string
 		hyperthreading        types.HyperthreadingMode
+		workloadCpuset        string
```
Change this to `[]types.WorkloadPartition` and insert it directly into the InstallConfig when creating it.

staebler commented on pkg/asset/machines/worker.go:
```diff
@@ -223,6 +223,14 @@ func (w *Worker) Generate(dependencies asset.Parents) error {
 		}
+		if len(ic.Workload) > 0 {
+			w := ic.Workload[0]
+			ignWorkload, err := machineconfig.ForWorkloadPartitioning(string(w.Name), w.CPUIds, "worker")
```
This gives me pause. I've asked a question in the enhancement PR about whether the same cpuset should be used for master and workers. See https://github.com/openshift/enhancements/pull/703#pullrequestreview-624640278.

lack replied:
Given that this will only be used for SNO in 4.8 it probably doesn't matter, but I agree it's worth designing this right out of the gate.

lack replied:
For this first release it probably doesn't matter, and if we want different pools in the future, perhaps it's not too hard to eventually also allow the same "WorkloadSettings" in install-config.yaml to be added under the `compute` and `controlPlane` in addition to it being globally available at the top level?

lack commented on pkg/types/installconfig.go:
```diff
@@ -351,3 +360,33 @@ type BootstrapInPlace struct {
+const (
+	// ManagementWorkloadPartition pins all infrastructure pods to the
+	// defined cpuset.
+	ManagementWorkloadPartition WorkloadPartitionName = "management"
```
That's an interesting thing... The explicit value `management` is what the name should be because then it will really match what the crio and kubelet configs get.  So do we really want the user input to not match what ends up in the configs?

staebler replied:
That makes sense. This isn't really an enum but rather a name with limited values.

lack commented on pkg/types/installconfig.go:
```diff
@@ -145,6 +145,15 @@ type InstallConfig struct {
+	// name must be "management", which will pin all infrastructure pods to
+	// the defined cpuset.
+	// +optional
+	Workload []WorkloadPartition `json:"workloadSettings,omitempty"`
```
Not as described today in the enhancement.

staebler replied:
I don't want to get locked in with having to add another field at the top level if we need to. I think I'd rather add one more level here now.

lack commented:
/retest

lack commented:
/retest

lack commented:
/retest

lack commented:
/retest

lack commented:
/test images

lack commented:
/retest

lack commented:
/retest

staebler commented on pkg/types/machinepools.go:
```diff
@@ -125,3 +132,29 @@ func (p *MachinePoolPlatform) Name() string {
+	// gives the user control in situations where they need specific CPUs
+	// to be available because they're close to accelerators, NICs, or
+	// other special hardware.  For example, "1,5,9" or "0-4" or
+	// "0-1,52-53" +kubebuilder:validation:Pattern=[0-9][-,0-9]*
```
```suggestion
	// "0-1,52-53"
	// +kubebuilder:validation:Pattern=[0-9][-,0-9]*
```

lack replied:
Good catch, not sure how that formatting got broken.

staebler commented on pkg/asset/machines/worker_test.go:
```diff
@@ -249,3 +256,59 @@ func TestComputeIsNotModified(t *testing.T) {
 	}
 }
+
+func TestWorkloadPartitioning(t *testing.T) {
```
What is this supposed to be testing? I don't see any mention of workloads in the function.

lack replied:
Another odd editing error, removing.

staebler commented on pkg/types/validation/installconfig.go:
```diff
@@ -395,6 +395,9 @@ func validateCompute(platform *types.Platform, control *types.MachinePool, pools
 		}
 		allErrs = append(allErrs, ValidateMachinePool(platform, &p, poolFldPath)...)
+		if len(p.Workloads) > 0 {
+			allErrs = append(allErrs, field.Invalid(poolFldPath.Child("workloads"), p.Workloads, "workloads can only be specified for controlPlane pools"))
```
```suggestion
			allErrs = append(allErrs, field.Invalid(poolFldPath.Child("workloads"), p.Workloads, "workloads can not be specified for compute pools"))
```

staebler commented on pkg/asset/machines/machineconfig/workload-partitioning.go:
```diff
@@ -0,0 +1,122 @@
+
+// ForWorkloads creates the MachineConfig that configures Cri-O and
+// Kubelet with the workload partition settings from install-config.yaml
+// 'workloadSettings'
```
What is this line meant to indicate?

lack replied:
A left-over from the bulk rename.  Fixing now.

staebler commented on pkg/asset/machines/worker.go:
```diff
@@ -223,6 +223,13 @@ func (w *Worker) Generate(dependencies asset.Parents) error {
 			}
 			machineConfigs = append(machineConfigs, ignFIPS)
 		}
+		if len(pool.Workloads) > 0 {
```
Don't put this code in here yet, since we don't support it.

lack replied:
Good point.

staebler commented on pkg/types/validation/installconfig.go:
```diff
@@ -395,6 +395,9 @@ func validateCompute(platform *types.Platform, control *types.MachinePool, pools
 			allErrs = append(allErrs, field.Invalid(poolFldPath.Child("architecture"), p.Architecture, "heteregeneous multi-arch is not supported; compute pool architecture must match control plane"))
 		}
 		allErrs = append(allErrs, ValidateMachinePool(platform, &p, poolFldPath)...)
+		if len(p.Workloads) > 0 {
```
Can we do this check in `ValidateMachinePool` instead so that we don't report potential errors for the specific workloads in addition to reporting that workloads as a whole are not allowed? You could either pass in the type of machine pool being validated or rely on the name of the machine pool to distinguish.

lack replied:
Good idea!  Moved.

staebler reviewed (commented):
This looks good to me overall. I have some nits. And there are some nits that I think may have been missed from the previous review (from github helpfully hiding them). I took the liberty of resolving the ones of those that are obsolete.

lack commented:
/retest

lack commented:
/retest

lack commented:
/retest

lack commented:
/retest

lack commented:
> This looks good to me overall. I have some nits. And there are some nits that I think may have been missed from the previous review (from github helpfully hiding them). I took the liberty of resolving the ones of those that are obsolete.

Latest push addresses all of those nits.  Thanks for all your time on this!

lack commented:
/retest

lack commented:
/retest

lack commented:
/retest

lack commented:
/retest

lack commented:
After some discussion today with @smarterclayton we're shelving this installer price of the feature for now.  Users opting in to this feature will manually create manifests (or use external tooling to do so.
